Switch to using WHATWG streams

pthatcherg opened the pull request:
First attempt at this.  It can probably use a lot of eyeballs.

pthatcherg commented on index.html:
```diff
@@ -94,30 +91,14 @@ <h2><dfn>UnidirectionalStreamsTransport</dfn> Mixin</h2>
 interface mixin UnidirectionalStreamsTransport {
   Promise&lt;SendStream&gt; createSendStream (optional SendStreamParameters parameters);
-  attribute EventHandler    onreceivestream;
+  ReadableStream            receiveStreams ();
```
It could, but that would preclude us from ever being able to add a parameters to it.  I'm not sure what parameters we would add, but we might come up with something.  Do you think an attribute would be much easier to use?

aboba replied:
Not sure I understand this.  Is the idea that `receiveStreams().getReader.read()` returns the most recently created `receiveStream`s?  Wouldn't an event be more efficient than having to poll?

pthatcherg replied:
Yes, and I changed the steps based on Steve's comments. Take a look at the latest.

pthatcherg commented on index.html:
```diff
@@ -193,6 +174,45 @@ <h2>Methods</h2>
+              on which <code>receiveStreams</code> is invoked.</p>
+            </li>
+            <li>
+              <p>If <code><var>transport</var>'s state</code> is <code>"closed"</code> or
```
Good idea.

pthatcherg commented on index.html:
```diff
@@ -345,29 +282,13 @@ <h2><dfn>BidirectionalStreamsTransport</dfn> Mixin</h2>
 interface BidirectionalStreamsTransport {
     Promise&lt;BidirectionalStream&gt; createBidirectionalStream ();
-    attribute EventHandler             onbidirectionalstream;
+    ReadableStream                     receiveBidirectionalStreams ();
```
Done

pthatcherg commented on index.html:
```diff
@@ -589,144 +465,58 @@ <h2>Attributes</h2>
+          <p>Return the value of the <dfn>[[\ReceivedDatagrams]]</dfn> internal slot.</p>
+          <p>For each datagram received, insert it into <dfn>[[\ReceivedDatagrams]]</dfn>.  
+             If too many datagrams are queued because the stream
+             is not being read quickly enough, datagrams to avoid queueing.<p>
```
Done

aboba commented:
@pthatcherg What specification should we reference for a stream of objects? While object streams are commonly supported in Javascript streams libraries (e.g. node.js, streams.js, etc.) I could not find mention of object streams in the WHATWG streams spec.

aboba commented:
Here is an example of the sockets API using streams: https://www.w3.org/2012/sysapps/tcp-udp-sockets/#interface-tcpsocket

pthatcherg commented:
Some replies to Bernard:

- ReceivedDatagrams and SentDatagrams are once for QuicTransport and once for Http3Transport.  If there's a way we can deduplicate that between them, I'd be happy to know it.  Do know of a way?
- I fixed the IncomingStream.readable attribute.  Thanks for catching that.
- I don't see those events any more.  Perhaps I removed them in a recent revision.
- You are correct about the examples and readyToSendDatagram.  I have removed those now.

yutakahirano commented:
cc: @ricea

aboba commented:
More nits: 

- Do we need to add a definition of `ReadableStream` to the terminology section, referencing [[WHATWG-STREAMS]]?  I see 8 link warnings to the missing dfn.
- [[\ReceiveStreams]] is defined at lines 190 and 195. Are either of these necessary?
. [[\ReceivedBidirectionalStreams]] is defined at lines 374 and 379.  Are either of these necessary?

ricea commented:
Looks good. There's less detail on the stream manipulation operations compared to, for example, the [Fetch standard](https://fetch.spec.whatwg.org/#readablestream), but maybe that's appropriate here.

I want to review the whole thing again in more detail as there was stuff I didn't understand without context. However, there's no need to block this on me; I can review it after it is landed.

pthatcherg commented:
In response to Bernard:

- I added a reference to https://streams.spec.whatwg.org.
- You're right about [[\RecievedStreams]] and [[\RecievedBidirectionalStreams]] and [[\ReceivedDatagrams]] and [[\SentDatagrams]].   I changed them to links and removed the extra "]".

steveanton commented:
I'm supportive of landing the PR soon (even with open issues) since it is a big change and touched a lot of the spec.

lgrahl commented on index.html:
```diff
@@ -589,144 +457,36 @@ <h2>Attributes</h2>
+          <p>Return the value of the <code><a>[[\ReceivedDatagrams]]</code></a> internal slot.</p>
+          <p>For each datagram received, insert it into <code><a>[[\ReceivedDatagrams]]</code></a>.
+             If too many datagrams are queued because the stream
+             is not being read quickly enough, drop datagrams to avoid queueing.<p>
```
How many datagrams should be queued? Which datagrams are being dropped and which are being inserted? Will one always get the most recent datagram or the first added to the queue? I think this is important to specify.

pthatcherg replied:
I don't think we can give an exact number of packets, but I added some text for guidance as to how to choose (not too big and not too small).

pthatcherg commented on index.html:
```diff
@@ -568,10 +437,9 @@ <h2><dfn>DatagramTransport</dfn> Mixin</h2>
-    Promise&lt;boolean&gt                     sendDatagram (Uint8Array data);
-    Promise&lt;sequence&lt;Uint8Array&gt;&gt; receiveDatagrams ();
+    readonly attribute unsigned short maxDatagramSize;
+    WritableStream                    sendDatagrams ();
```
It's of Uint8Array.  I added some text to make that more clear.

pthatcherg commented:
In response to Lennart: 

- The rendering was busted because I forgot a "," in the config.  It's fixed now. 

- I don't think we can use ReadableStream<T> and WritableStream<T>.  I wish we could.  Victor asked around about this and he said the right answer is to explain it in the spec text, not the WebIDL, which is what I've done.

- I don't think it makes sense to use the body mixin for datagrams because they aren't streams.  They are an array of bytes, so the type for array of bytes makes sense.  It also doesn't make sense for UnidirectionalStreams or BidirectionalStreams either because those are more than just a payload/body/stream-of-bytes.  And I don't think it makes sense to have the .readable of those be a body mixin instead of just being a ReadableStream.  Then the poor user of the API has to deal with a bunch of things it might be instead of just one thing.

- I added a note explaining the distinction between "streams" here and "streams" in WHATWG streams.

pthatcherg commented:
Given Steve's comment, I will merge this PR by the end of the week (around 24 hours from now) unless someone brings up any new blocking issues or asks for more time.

aboba commented:
Am still seeing 2 respec errors, for duplicate definitions of [[\SentDataGrams]] and [[\ReceivedDataGrams]].

pthatcherg commented:
Yeah, that's because it's defined in two different classes.  I don't know how to fix it :).

pthatcherg commented:
It's merged.  I hope everyone is OK with that :).  We can fix up things in more PRs.
